File: csharp/Magpie.Compilation/Parser/Token.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Magpie.Compilation
{
    public enum TokenType
    {
        LeftParen,
        RightParen,
        LeftBracket,
        RightBracket,
        RightBracketBang, // ]! used for mutable arrays
        LeftCurly,
        RightCurly,
        LeftArrow,
        RightArrow,
        Comma,
        Colon,
        Dot,
        Line,
        Prime, // '

        // keywords
        Case,
        Def,
        Do,
        Else,
        End,
        Fn,
        For,
        If,
        Let,
        Match,
        Namespace,
        Return,
        Struct,
        Then,
        Union,
        Using,
        Var,
        While,
        
        // literals
        Bool,
        Int,
        String,
        
        // identifiers
        Name,
        Operator,
        
        // the end of the source has been reached
        Eof
    }

    public class Token
    {
        public Position Position;
        public TokenType Type;
        public bool BoolValue;
        public int IntValue;
        public string StringValue;

        public Token(Position position, TokenType type)
            : this(position)
        {
            Type = type;
        }

        public Token(Position position, TokenType type, string text)
            : this(position)
        {
            Type = type;
            StringValue = text;
        }

        public Token(Position position, bool value)
            : this(position, TokenType.Bool)
        {
            BoolValue = value;
        }

        public Token(Position position, int value)
            : this(position, TokenType.Int)
        {
            IntValue = value;
        }

        public Token(Position position, string value)
            : this(position, TokenType.String)
        {
            StringValue = value;
        }

        private Token(Position position)
        {
            Position = position;
        }

        public override string ToString()
        {
            switch (Type)
            {
                case TokenType.LeftParen:     return "(";
                case TokenType.RightParen:    return ")";
                case TokenType.LeftBracket:   return "[";
                case TokenType.RightBracket:  return "]";
                case TokenType.RightBracketBang:  return "]!";
                case TokenType.LeftCurly:     return "{";
                case TokenType.RightCurly:    return "}";
                case TokenType.LeftArrow:     return "<-";
                case TokenType.RightArrow:    return "->";
                case TokenType.Comma:         return ",";
                case TokenType.Colon:         return ":";
                case TokenType.Dot:           return ".";
                case TokenType.Line:          return "newline";
                
                case TokenType.Name:          return "name " + StringValue;
                case TokenType.Operator:      return "operator " + StringValue;

                case TokenType.Case:        return "case";
                case TokenType.Def:         return "def";
                case TokenType.Do:          return "do";
                case TokenType.Else:        return "else";
                case TokenType.End:         return "end";
                case TokenType.Fn:          return "fn";
                case TokenType.For:         return "for";
                case TokenType.If:          return "if";
                case TokenType.Let:         return "let";
                case TokenType.Match:       return "match";
                case TokenType.Namespace:   return "namespace";
                case TokenType.Struct:      return "struct";
                case TokenType.Then:        return "then";
                case TokenType.Union:       return "union";
                case TokenType.Using:       return "using";
                case TokenType.Var:         return "var";
                case TokenType.While:       return "while";

                case TokenType.Bool:        return "bool " + BoolValue;
                case TokenType.Int:         return "int " + IntValue;
                case TokenType.String:      return "string " + StringValue;

                case TokenType.Eof:         return "[end]";
                default: throw new Exception();
            }
        }
    }
}

File: src/com/stuffwithstuff/magpie/Token.java
package com.stuffwithstuff.magpie;

import com.stuffwithstuff.magpie.TokenType;
import com.stuffwithstuff.magpie.ast.Position;


public final class Token {
  public Token(Position position, TokenType type) {
    mPosition = position;
    mType = type;
    mValue = null;
  }

  public Token(Position position, TokenType type, Object value) {
    mPosition = position;
    mType = type;
    mValue = value;
  }

  public Position getPosition() { return mPosition; }
  
  public TokenType getType() { return mType; }
  
  public boolean getBool()   { return ((Boolean)mValue).booleanValue(); }
  public int     getInt()    { return ((Integer)mValue).intValue(); }
  public double  getDouble() { return ((Double)mValue).doubleValue(); }
  public String  getString() { return (String)mValue; }
  
  public String toString() {
    switch (mType)
    {
      case LEFT_PAREN: return "(";
      case RIGHT_PAREN: return ")";
      case LEFT_BRACKET: return "[";
      case RIGHT_BRACKET: return "]";
      case LEFT_BRACE: return "{";
      case RIGHT_BRACE: return "}";
      case COMMA: return ",";
      case LINE: return "(line)";
      case DOT: return ".";
      case EQUALS: return "=";

      case NAME: return getString() + " (name)";
      case OPERATOR: return getString() + " (op)";

      case BOOL: return Boolean.toString(getBool());
      case INT: return Integer.toString(getInt());
      case DOUBLE: return Double.toString(getDouble());
      case STRING: return "\"" + getString() + "\"";

      case ARROW: return "->";
      case CASE: return "case";
      case CLASS: return "class";
      case DEF: return "def";
      case DO: return "do";
      case ELSE: return "else";
      case END: return "end";
      case EXTEND: return "extend";
      case FOR: return "for";
      case FN: return "fn";
      case IF: return "if";
      case LET: return "let";
      case MATCH: return "match";
      case SHARED: return "shared";
      case THEN: return "then";
      case THIS: return "this";
      case VAR: return "var";
      case WHILE: return "while";

      case EOF: return "(eof)";

      default: return "(unknown token?!)";
    }
  }
  
  private final Position  mPosition;
  private final TokenType mType;
  private final Object    mValue;
}
